test/distributed/fsdp/test_fsdp_sharded_grad_scaler.py

```diff
@@ -0,0 +1,153 @@
+        s = torch.sparse_coo_tensor(i, v, torch.Size([2, 3]), device="cpu", dtype=torch.float)
+
+        # unscale sparse tensors
+        s1 = s.clone()
+        s1.grad = s.clone()
+        opt = torch.optim.SGD([s1], lr=1.0)
+        found_inf.zero_()
+        found_inf = scaler._unscale_grads_(opt, inv_scale, found_inf)[s1.device]
+        self.assertEqual(found_inf, 0.0)
+        self.assertEqual(s1.grad.to_dense(), (s / 2).to_dense())
+
+        # unscale sparse tensor: inf
+        v = torch.tensor([16.0, 32.0, float('inf')], dtype=torch.float, device="cpu")
+        s1.grad = torch.sparse_coo_tensor(i, v, torch.Size([2, 3]), device="cpu", dtype=torch.float)
+        found_inf.zero_()
+        found_inf = scaler._unscale_grads_(opt, inv_scale, found_inf)[s1.device]
+        self.assertEqual(found_inf, 1.0)
+
+        # unscale sparse tensor: overflow (marked as inf)
+        i = torch.tensor([[1, 1, 1],
+                          [0, 0, 2]], device="cpu", dtype=torch.int64)
+        # coalescing sparse tensor here will cause the value to be Inf
+        v = torch.tensor([2**15, 2**15, 1.0], dtype=torch.float16, device="cpu")
+        s1 = torch.sparse_coo_tensor(i, v, torch.Size([2, 3]), device="cpu", dtype=torch.float16)
+        s1.grad = s1.clone()
+        found_inf.zero_()
+        found_inf = scaler._unscale_grads_(opt, inv_scale, found_inf)[s1.device]
+        self.assertEqual(found_inf, 1.0)
+
+    def test_inf_gradients_skip_optim_step(self):
+        pg = DummyProcessGroup(0, 1)
+        scaler = ShardedGradScaler(init_scale=2.0, process_group=pg, enabled=True)
+        loss = torch.full((1,), 4.0, dtype=torch.float32, device="cpu")
+        t0 = torch.tensor([float('inf')], dtype=torch.float32, device="cpu")
+        t0.grad = t0.clone()
+        opt = torch.optim.SGD([t0], lr=1.0)
+        scaler.scale(loss)
+        ret_val = scaler.step(opt)
+        self.assertTrue(ret_val is None)
+
+
+class TestShardedGradScalerParityWithDDP(FSDPTest):
+    def _get_init_modes_for_test(self, cpu_offload):
+        modes = [
+            FSDPInitMode.CUDA_AFTER,
+            FSDPInitMode.CUDA_BEFORE
+        ]
+        # Note that FSDPInitMode.CUDA_NEVER works currently only with CPU
+        # offload as we explicitly bring the param back to CUDA device. In
+        # general, it will not work since we try to all_gather p.data which is
+        # on CPU but NCCL only supports GPU.
+        if cpu_offload.offload_params:
+            modes.append(FSDPInitMode.CUDA_NEVER)
+
+        return modes
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(params, configs, subtest_name)
+    def test_scaler_enabled(self, cpu_offload, sharding_strategy, mixed_precision):
+        init_modes = self._get_init_modes_for_test(cpu_offload)
+        mp = MixedPrecision(
+            param_dtype=torch.float16,
+            reduce_dtype=torch.float16,
+            buffer_dtype=torch.float16,
```

Comment:
As a follow up, can we also add bfloat16 tests. bfloat16 is hardware conditional, there are some logic in test_fsdp_mixed_precision to make sure we only run it where appropriate.

Reply:
bfloat16 is tested in test_mixed_precision.py above.